### DISCUSSION TITLE
Browse Reddit Directly in Script Kit

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#177] [Share]
https://user-images.githubusercontent.com/36073/115759214-3b78f180-a35d-11eb-9c96-2af22c268803.mp4

[Install reddit](https://scriptkit.app/api/new?name=reddit&url=https://gist.githubusercontent.com/johnlindquist/05c75efeebd2c2ca0149ee4e92aeb58d/raw/4ad69ed25572d5527db84f2db5c578ad0aacde25/reddit.js)

```js
//Shortcut: opt r

/**
 * @typedef {typeof import("reddit")} Reddit
 */

/*
First, create a Reddit App.

Click "Create app"
For simple scripts, you can select a type of "script".
You can enter anything in the "about URL" and "redirect URL" fields.
Note your app ID (appears below the app name) and your app secret.
*/

/** @type Reddit */
let Reddit = await npm("reddit")

let envOptions = {
  ignoreBlur: true,
  hint: md(
    `[Create a reddit app](https://www.reddit.com/prefs/apps)`
  ),
  secret: true,
}

let reddit = new Reddit({
  username: await env("REDDIT_USERNAME"),
  password: await env("REDDIT_PASSWORD"),
  appId: await env("REDDIT_APP_ID", envOptions),
  appSecret: await env("REDDIT_APP_SECRET", envOptions),
  userAgent: `ScriptKit/1.0.0 (https://scriptkit.com)`,
})

let subreddits = [
  "funny",
  "aww",
  "dataisbeautiful",
  "mildlyinteresting",
  "RocketLeague",
]

subreddits.forEach(sub => {
  onTab(sub, async () => {
    await arg("Select post to open:", async () => {
      let best = await reddit.get(`/r/${sub}/hot`)

      return best.data.children.map(({ data }) => {
        let {
          title,
          thumbnail,
          url,
          subreddit_name_prefixed,
          preview,
        } = data

        let resolutions = preview?.images?.[0]?.resolutions
        let previewImage =
          resolutions?.[resolutions?.length - 1]?.url

        return {
          name: title,
          description: subreddit_name_prefixed,
          value: url,
          img: thumbnail,
          ...(previewImage && {
            preview: md(`
![${title}](${previewImage})

### ${title}          
                `),
          }),
        }
      })
    })
  })
})

```

**Comment by craigz** [comment]
having added this script and modifying the subreddits to taste, upon execution the tab bar appears for a brief moment and the entire window closes before i have a chance to get my cursor to select a tab. is there a way to make the script pause/await input before closing?

**Reply by johnlindquist** [reply]
Any chance you could share the specific changes you made? There might be an error or something causing that behavior.

**Reply by craigz** [reply]
this [gist](https://gist.github.com/craigz/7ab0ff8e456dd4ead84732b6a7b96bb6) has the `reddit.js` script that i'm using which is a direct copy of this [post](https://github.com/johnlindquist/kit/discussions/177). 

the script as is exhibits the described behavior. the change i had made was to the `subreddits` array which for the sake of example i've returned to the sample reddits in the gist.

thanks for checking this out 👍

**Reply by craigz** [reply]
additionally, i don't think there is an issue with the reddit app/authentication, and the scriptkit log has this to say:
```log
13487: SET_SCRIPT Prompt process reddit.js
[2022-02-03 00:57:15.815] [info]  13487: SET_TAB_INDEX Prompt process reddit.js
[2022-02-03 00:57:15.816] [info]  13487: SET_PROMPT_DATA Prompt process reddit.js
[2022-02-03 00:57:16.488] [info]  13487: 🟡 exit 1. Prompt process: /Users/craigz/.kenv/scripts/reddit.js
[2022-02-03 00:57:16.489] [info]  13487: 🛑 removed
```
not so much in the way of errors, just the window flashes and vanishes after drawing the tab bar, listing subreddits.

**Reply by craigz** [reply]
btw, i've returned to this script, and in testing, i've installed a fresh copy of the script from https://www.scriptkit.com/johnlindquist/reddit and the identical behavior is occurring.  
- the tab bar is loaded, listing the subreddits listed in the array as tab bar items.
- the scriptkit window immediately disappears.

is there anything i can do to make the `onTab` items stay in place until i can access them? this is the output of the example reddit script unchanged, and running in real time.
![scriptkit_reddit](https://user-images.githubusercontent.com/1929275/158047449-45cc5906-7545-4dab-8804-d25744f2e831.gif)

**Reply by johnlindquist** [reply]
Sorry, I'm having trouble understanding the issue:

The issue is stated as "scriptkit window immediately disappears", but that gif is flashing between the tabs loading then returning to the script selection UI. Is it supposed to be showing the UI disappearing?

The only thing I can think of is that the reddit app id/credentials are incorrect. Here's what a fresh version of the script looks like on my machine:

https://user-images.githubusercontent.com/36073/158203098-84ae5c90-4154-4406-8ec2-317a1e6934af.mp4

**Reply by craigz** [reply]
Strange, it looks much better on yours than mine! At least it shows that it is possible for the script to work, that'll keep me running into this wall for a bit longer ;)

In my case, as soon as i run the script and the menu bar is drawn, it simply closes the entire script kit window. The video appears slightly different because it's looping from the selection and not showing the disappearing window.  As far as I can tell the reddit app credentials are ok, I will look into another way of testing that the app credentials work, and if that proves difficult will simply create a brand new reddit app. I'm not seeing any immediate errors regarding authentication or anything, and scriptkit is not logging any specific errors. 

Thanks again for looking into this.
